app/helpers/react_on_rails_helper.rb

```diff
@@ -2,132 +2,196 @@
+    dom_id = "#{component_name}-react-component-#{react_component_index}"
+
+    # Setup the page_loaded_js, which is the same regardless of prerendering or not!
+    # The reason is that React is smart about not doing extra work if the server rendering did its job.
+    data_variable_name = "__#{component_name.camelize(:lower)}Data#{react_component_index}__"
     turbolinks_loaded = Object.const_defined?(:Turbolinks)
-    install_render_events = turbolinks_loaded ? turbolinks_bootstrap(domId) : non_turbolinks_bootstrap
-
+    install_render_events = turbolinks_loaded ? turbolinks_bootstrap(dom_id) : non_turbolinks_bootstrap
     page_loaded_js = <<-JS
-      (function() {
-        window.#{dataVariable} = #{props.to_json};
-        #{define_render_if_dom_node_present(reactComponent, dataVariable, domId, trace)}
-        #{install_render_events}
-      })();
+(function() {
+  window.#{data_variable_name} = #{props.to_json};
+  #{define_render_if_dom_node_present(react_component_name, data_variable_name, dom_id, trace(options), generator_function(options))}
+  #{install_render_events}
+})();
     JS
 
     data_from_server_script_tag = javascript_tag(page_loaded_js)
 
     # Create the HTML rendering part
-    if prerender
+    server_rendered_html =
+      server_rendered_react_component_html(options, props, react_component_name)
+
+    rendered_output = content_tag(:div,
+                                  server_rendered_html,
+                                  id: dom_id)
+
+    # IMPORTANT: Ensure that we mark string as html_safe to avoid escaping.
+    <<-HTML.html_safe
+#{data_from_server_script_tag}
+#{rendered_output}
+    HTML
+  end
+
+  def next_react_component_index
+    @react_component_index ||= -1
+    @react_component_index += 1
+  end
+
+  def server_rendered_react_component_html(options, props, react_component_name)
+    if prerender(options)
       render_js_expression = <<-JS
-          renderReactComponent(this.#{reactComponent}, #{props.to_json})
+        (function(React) {
+          var reactElement = #{render_js_react_element(react_component_name, props.to_json, generator_function(options))};
+          return React.renderToString(reactElement);
+        })(this.React);
       JS
-      server_rendered_react_component_html = render_js(render_js_expression)
+      # create the server generated html of the react component with props
+      options[:react_component_name] = react_component_name
+      server_rendered_react_component_html =
+        render_js(render_js_expression, options)
     else
       server_rendered_react_component_html = ""
     end
-
-    rendered_output = content_tag(:div,
-                                  server_rendered_react_component_html,
-                                  id: domId)
-
-    <<-HTML.strip_heredoc.html_safe
-      #{data_from_server_script_tag}
-      #{rendered_output}
-    HTML
+    server_rendered_react_component_html
```

Comment:
that is correct, and if the return is the last statement, then we don't need the return either.